theano/tensor/sort.py

```diff
@@ -206,3 +219,311 @@ def argsort(a, axis=-1, kind='quicksort', order=None):
+            op=self.__class__.__name__, axis=self.axis)
+
+    def make_node(self, inp, kth):
+        # numpy always uses int64 as output dtype for arg*() routines
+        # however, we add this option as memory is more precious on gpu
+        inp = theano.tensor.as_tensor_variable(inp)
+        ndim = inp.ndim
+        if ndim == 0:
+            raise ValueError('Cannot take scalar as input')
+        if not -ndim <= self.axis < ndim:
+            raise IndexError(
+                '"axis" parameter out of range,'
+                ' expected integer within [%d, %d]' % (-ndim, ndim - 1))
+
+        kth = theano.tensor.as_tensor_variable(kth)
+        bcast = inp.type.broadcastable
+        outs = []
+        if self.return_values:
+            outs.append(inp.type())
+        if self.return_indices:
+            outs.append(theano.tensor.TensorType(
+                dtype=self.idx_dtype, broadcastable=bcast)())
+        return theano.Apply(self, [inp, kth], outs)
+
+    def perform(self, node, inputs, output_storage):
+        x, k = inputs
+        axis = self.axis
+        if not self.return_indices:
+            pzv = output_storage[0]
+            pzv[0] = _topk_py_impl(self, x, k, axis, None)
+        elif self.return_values:
+            pzv = output_storage[0]
+            pzi = output_storage[1]
+            pzv[0], pzi[0] = _topk_py_impl(
+                self, x, k, axis, node.outputs[1].dtype)
+        else:
+            pzi = output_storage[0]
+            pzi[0] = _topk_py_impl(self, x, k, axis, node.outputs[0].dtype)
+
+    def infer_shape(self, node, inp_shapes):
+        _check_tensor_is_scalar(node.inputs[1])
+        shp = list(inp_shapes[0])
+        shp[self.axis] = np.abs(node.inputs[1])
+        shp = tuple(shp)
+        return [shp for i in [self.return_values, self.return_indices] if i]
+
+    def L_op(self, inputs, outputs, out_grads):
+        x, k = inputs
+        k_grad = grad_undefined(self, 1, k, 'topk: k is not differentiable')
+        if not (self.return_indices, self.return_values):
+            x_grad = grad_undefined(
+                self, 0, x, 'topk: cannot get gradient'
+                ' without both indices and values')
+        elif x.ndim == 1:
+            z_grad = out_grads[0]
+            indices = outputs[-1]
+            x_grad = x.zeros_like(dtype=z_grad.dtype)
+            x_grad = theano.tensor.advanced_set_subtensor1(
+                x_grad, z_grad, indices)
+        else:
+            x_shp = theano.tensor.shape(x)
+            z_grad = out_grads[0]
+            ndim = x.ndim
+            axis = self.axis % ndim
+            grad_indices = [
+                arange(x_shp[i]).dimshuffle([0] + ['x'] * (ndim - i - 1))
+                if i != axis else outputs[-1] for i in range(ndim)]
+            x_grad = x.zeros_like(dtype=z_grad.dtype)
+            x_grad = theano.tensor.advanced_set_subtensor(
+                x_grad, z_grad, *grad_indices)
```

Comment:
I think this case can be simplified and can give more optimized code like this:

x_grad = x.zeros_like(dtype=z_grad.dtype)
grad_indices = [slice(None) if i != axis else outputs[-1] for i in range(ndim)]
x_grad = set_subtensor(x_grad[*grad_indices], z_grad, )

Why this would be more optimized? Theano will sometimes reuse advancedincsubtensor1 in such case as it is faster. Your version force the advancedincsubtensor that isn't as much optimized.